<!--
    index.html ~~
                                                    ~~ (c) JSA and SRW, 2012
-->
<html>
  <head>
    <script src="http://qmachine.org/q.js"></script>
    <script src="fasta.js"></script>
    <link rel="stylesheet" type="text/css" href="style.css" />
  </head>
  <body>
    <noscript>This page requires JavaScript.</noscript>
    <h1>CGR Demo for QMachine</h1>
    <div>
      Experimenting with using QMachine (QM) for analysis of multiple
      Prokaryote genomes: http://q.cgr.googlecode.com/hg/index.html, see
      <a href="#">YouTube</a> for a webcast.
    </div>
    <div>
      In this experiment we will look for the longest similar segment in the
      fourteen <i>Streptococcus pneumoniae</i> strains for which the full
      genome is available at NCBI's FTP site as of April 30, 2012.
    </div>
    <div>
      The first step is to make QMachine's JavaScript library available, as
      shown in the head of this document:
    </div>
    <pre>
&lt;script src="http://qmachine.org/q.js"&gt;&lt;/script&gt;
    </pre>
    <div>
      It could also have been imported programmatically by:
    </div>
    <pre>
s=document.createElement('script');
s.src='http://qmachine.org/q.js';
document.body.appendChild(s);
    </pre>
    <div>
      In this illustration we will distribute the mapReduce decomposition of
      sequence alignment described in (Almeida 2012),
      <a href="http://usm.github.com">http://usm.github.com</a>. In that
      report, the genome of one of those strains, R6, is probed for the longest
      similar segment to "TCCACAGCATGCGTGACGATGACACG"
    </div>
    <pre>
uBac = new usm('ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Streptococcus_pneumoniae_R6_uid57859/NC_003098.fna')
A = uBac.align('TCCACAGCATGCGTGACGATGACACG')
    </pre>
    <div>
      and three different segments of ten units each are found. Here, the
      fourteen genomes, each approximately two million base-pairs in length,
      are also made available as JavaScript function calls at
      <a href="//q.cgr.googlecode.com/hg/Pneumo"
        target=_blank>http://q.cgr.googlecode.com/hg/Pneumo</a>.
      Using the R6 strain as an example, the
      <a href="//q.cgr.googlecode.com/hg/Pneumo/NC_003098.fna.js">fasta()</a>
      call looks like this:
    </div>
    <pre>
fasta({name:"NC_003098.fna",
head:">gi|15902044|ref|NC_003098.1| Streptococcus pneumoniae R6, complete genome",
body:"TTGAAAGAAAAACAATTTTGGAATCGTATATTAGAATTTGCACAAGAAAGACTGACTCGATCCATGTATGATTTCTA...TGCTATCTATGGTAAAATATCTCTAGT"})
    </pre>
    <div>
      The function <a href="fasta.js">fasta()</a> will now be defined such that
      the computation amy be distributed by QM for execution by other web
      browsers in parallel where this sequence matching will be made for each
      of the fourteen genomes and the results returned back to the invoking
      machine:
    </div>
    <pre>
	fasta=function(x){
		if(!!x){fasta.seq=x}; // store sequence in fasta() 
		// Get the USM library and its dependencies if it is not there already
		if(typeof(usm)=='undefined'){
			//fasta.load(['http://localhost:8888/jmat/jmat.js','http://localhost:8888/usm/usm.js'],fasta)
			fasta.load(['jmat.googlecode.com/git/jmat.js'],['http://usm.github.com/usm.js'])
			}
		else{
			console.log('Indexing genome ...')
			//var uBac = new usm(x.body); // for small genomes this would be enough
			var uBac = new usm();
			uBac.encodeLong(fasta.seq.body,'ACGT');
			console.log('Aligning new sequence to indexed genome...')
			fasta.A = uBac.align('TCCACAGCATGCGTGACGATGACACG'); // store result in fasta.A
			console.log('done !');
			// Sean, I am guessing in this position you want to have something returning fasta.A to the Q
		}
	}

	fasta.load=function(url,callback){ // to load fasta call from remote location
		if(typeof(url)=='string'){url=[url,'']} // to make sure it's an array of strings, there may be more than one
		else if(url[url.length-1]!==''){url.push('')} // making sure trailing '' is there, it will be used to close iteration
		if(url[0].length>0){
			console.log('loading '+url[0]+' ...');
			var s=document.createElement('script');
			s.src=url[0];
			s.onload=function(){
				fasta.load(url.slice(1),callback)
				}
			document.body.appendChild(s);	
		}
		else if(!!callback){callback()}
	}
    </pre>
    <div>
      For convenience, this function was also loaded in the header of this
      document, so we are already ready to use it. The fasta function includes
      a method that will help us load the pneumococcal genomes and trigger the
      main function, fasta(), to align the test segment to it. For example, for
      the R6 strain we could do this by
    </div>
    <pre>
fasta.load('http://q.cgr.googlecode.com/hg/Pneumo/NC_003098.fna.js')
    </pre>
    <div>
      This will take a few seconds -- just wait for the console to display
      "done!". The final result is stored in fasta.A, which contains the
      position or positions of the longest shared segments and will be
      something like,
    </div>
    <pre>
	fasta.A
	Object
	ind: 1
	match: Array[4]
	0: 0
	1: 10
	2: 10
	3: 10
	length: 4
	__proto__: Array[0]
	posBase: Array[4]
	0: 0
	1: 1811967
	2: 1895547
	3: 1992091
	length: 4
	__proto__: Array[0]
	posProbe: Array[4]
	0: 0
	1: 5
	2: 0
	3: 14
	length: 4
	__proto__: Array[0]
	__proto__: Object
    </pre>
    <div>
      As detailed in the USM paper, the longest shared segment has length 10
      and 10 different segments of that length were found in R6's genome. Now,
      for the grand finale, let's distribute the 14 pneumococcal genomes to
      volunteer machines using QM to find out what would be the alignment
      result for each of them. As detailed in the QMachine manuscript, this
      will distribute both the computational resources and the bandwidth:
    </div>
    <pre>
Your turn Sean, pls make it as simple as possible, note we only need to retrieve the fasta.A results from each alignment.
    </pre>
    <div>
      <a href="user.html">User Page</a>
      <a href="volunteer.html">Volunteer Page</a>
    </div>
  </body>
</html>
