<!DOCTYPE html>
<!--
    index.html ~~
                                                    ~~ (c) JSA and SRW, 2012
-->
<html lang="en">
  <head>
    <meta charset="utf-8"/>
    <title>CGR Demo for QMachine</title>
    <link rel="stylesheet" href="./style.css"/>
    <script>
      (function () {
          'use strict';
       // This function immediately redirects the browser to a new URL that
       // has query parameters explicitly specified in order to demonstrate
       // their analogy to command-line arguments for the example daemon.
          if (location.search.slice(1) === '') {
              location.search = 'fasta-demo';
          }
          return;
      }());
    </script>
    <script src="http://qmachine.org/q.js"></script>
    <script src="./fasta.js"></script>
  </head>
  <body>
    <noscript>This page requires JavaScript.</noscript>
    <h1>CGR Demonstration for QMachine</h1>
    <div>
      Experimenting with using QMachine (QM) for analysis of multiple
      Prokaryote genomes: http://q.cgr.googlecode.com/hg/index.html, see
      <a href="#">YouTube</a> for a webcast.
    </div>
    <div>
      In this experiment we will look for the longest similar segment in the
      fourteen <i>Streptococcus pneumoniae</i> strains for which the full
      genome is available at
      <a href="ftp://ftp.ncbi.nih.gov/genomes/Bacteria/">NCBI's FTP site</a>
      as of April 30, 2012.
    </div>
    <div>
      The first step is to make QMachine's JavaScript library available, as
      shown in the head of this document:
    </div>
    <div class="inline-code">
      &lt;script src="http://qmachine.org/q.js"&gt;&lt;/script&gt;
    </div>
    <div>
      It could also have been imported programmatically:
    </div>
    <div class="inline-code">
      s = document.createElement('script');
      s.src = 'http://qmachine.org/q.js';
      document.body.appendChild(s);
    </div>
    <div>
      In this illustration we will distribute the MapReduce decomposition of
      sequence alignment described in (Almeida 2012),
      <a href="http://usm.github.com">http://usm.github.com</a>. In that
      report, the genome of one of those strains, R6, is probed for the longest
      similar segment to "TCCACAGCATGCGTGACGATGACACG",
    </div>
    <div class="inline-code">
      uBac = new usm('ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria/Streptococ\
        cus_pneumoniae_R6_uid57859/NC_003098.fna');
      A = uBac.align('TCCACAGCATGCGTGACGATGACACG');
    </div>
    <div>
      and three different segments of ten units each are found. Here, the
      fourteen genomes, each approximately two million base pairs long, are
      also made available as JavaScript function calls at
      <a href="Pneumo">http://q.cgr.googlecode.com/hg/Pneumo</a>. An
      abbreviated form of the
      <a href="Pneumo/NC_003098.fna.js">
        <span class="inline-code">fasta</span> call for R6</a> follows:
    </div>
    <div class="inline-code">
      fasta({
          name:"NC_003098.fna",
          head:">gi|15902044|ref|NC_003098.1| Streptococcus pneumoniae R6, c\
            omplete genome",
          body:"TTGAAAGAAAAACAATTTTGGAATCGTATATTAGAA...TATTAAACTTTTAAATAGTAA\
            AGGAGGAGAAAGGA"
      });
    </div>
    <div>
      The function <a class="inline-code" href="fasta.js">fasta</a> will now
      be defined such that the computation may be distributed by QM for
      execution by other web browsers in parallel where this sequence matching
      will be made for each of the fourteen genomes and the results returned
      back to the invoking machine:
    </div>
    <div class="inline-code" style="color:red">
      fasta = function (x) {
       // This function will be pasted in here as soon as I can figure out
       // which function Jonas wanted to include here. Then, I will probably
       // use an AJAX call inside the "markup.js" script to put the correct
       // definition here verbatim, just in case we edit or update it later.
          var y;
          // ...
          return y;
      };
    </div>
    <div>
      For convenience, this function was also loaded in the header of this
      document, so we are already ready to use it. The
      <span class="inline-code">fasta</span> function includes a method that
      will help us load the pneumococcal genomes and trigger the main function,
      <span class="inline-code">fasta()</span>, to align the test segment to
      it. For example, for the R6 strain we can use this invocation:
    </div>
    <div class="inline-code">
      fasta.load('http://q.cgr.googlecode.com/hg/Pneumo/NC_003098.fna.js');
    </div>
    <div>
      This will take a few seconds -- just wait for the console to display
      "done!". The final result is stored in
      <span class="inline-code">fasta.A</span>, which contains
      the position or positions of the longest shared segments and looks
      something like
    </div>
    <div class="inline-code">
      fasta.A
	Object
	ind: 1
	match: Array[4]
	0: 0
	1: 10
	2: 10
	3: 10
	length: 4
	__proto__: Array[0]
	posBase: Array[4]
	0: 0
	1: 1811967
	2: 1895547
	3: 1992091
	length: 4
	__proto__: Array[0]
	posProbe: Array[4]
	0: 0
	1: 5
	2: 0
	3: 14
	length: 4
	__proto__: Array[0]
	__proto__: Object
    </div>
    <div>
      As detailed in the USM paper, the longest shared segment has length 10,
      and 10 different segments of that length were found in R6's genome. Now,
      for the grand finale, let's compute the the alignment results for each
      of the 14 pneumococcal genomes simultaneously by using QM to distribute
      individual tasks to volunteer web browsers. As detailed in the QM
      manuscript, this strategy not only distributes the computational load,
      but it also distributes the bandwidth requirements, too.
    </div>
    <div style="color:red">
      (placeholder)
    </div>
    <div>
      <button onclick="Q.lib('./fasta-demo.js')">Launch job on QM</button>
      <button onclick="Q.lib('./daemon.js')">Volunteer for QM</button>
    </div>
    <div>
      Note also that, thanks to the use of the
      <a href="http://www.w3.org/TR/cors/">Cross-Origin Resource Sharing</a>
      standard that is available in modern web browsers, you can construct
      simple web pages to send to others so that, by simple opening the page,
      they can volunteer their computers' resources to help run your jobs.
      As an example, see
      <a href="./volunteer.html?fasta-demo">volunteer.html</a>.
    </div>
    <script src="markup.js"></script>
  </body>
</html>
